File: ENCOUNTERS-MS/service/encounter-service.go
package service

import (
	"ENCOUNTERS-MS/model"
	"ENCOUNTERS-MS/repo"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
)

type EncounterService struct {
	EncounterRepo *repo.EncounterRepository
}

func (service *EncounterService) GetApproved() ([]*model.Encounter, error) {

	if encounters, err := service.EncounterRepo.GetApproved(); err == nil {
		return encounters, nil
	}

	return nil, fmt.Errorf("Couldn't find any.")

}

func (service *EncounterService) GetApprovedByStatus(status model.EncounterStatus) ([]*model.Encounter, error) {
	if encounters, err := service.EncounterRepo.GetApprovedByStatus(status); err == nil {
		return encounters, nil
	}
	return nil, fmt.Errorf("Couldn't find any.")

}

func (service *EncounterService) GetByUser(userId int) ([]*model.Encounter, error) {
	if encounters, err := service.EncounterRepo.GetByUser(userId); err == nil {
		return encounters, nil
	}
	return nil, fmt.Errorf("Couldn't find any.")
}

func (service *EncounterService) GetTouristCreatedEncounters() ([]*model.Encounter, error) {
	if encounters, err := service.EncounterRepo.GetTouristCreatedEncounters(); err == nil {
		return encounters, nil
	}
	return nil, fmt.Errorf("Couldn't find any.")
}

func (service *EncounterService) GetNearbyHidden(userId int) ([]*model.Encounter, error) {

	lat, long, err := getLatLong(userId)

	if err != nil {
		return nil, err
	}

	if encounters, err := service.EncounterRepo.GetByType(long, lat, model.EncounterType("LOCATION")); err == nil {
		encounters = findInVicinity(&encounters, lat, long)
		return encounters, nil
	}
	return nil, fmt.Errorf("Couldn't find any.")
}

func (service *EncounterService) GetNearby(userId int) ([]*model.Encounter, error) {
	lat, long, err := getLatLong(userId)
	if err != nil {
		return nil, err
	}

	if encounters, err := service.EncounterRepo.GetAll(); err == nil {

		encounters = findInVicinity(&encounters, lat, long)

		return encounters, nil
	}

	return nil, fmt.Errorf("Couldn't find any.")
}

func (service *EncounterService) Create(encounter *model.Encounter) (*model.Encounter, error) {

	if res, err := service.EncounterRepo.Create(encounter); err == nil {
		return res, nil
	}
	return nil, fmt.Errorf("Couldn't create.")

}

func (service *EncounterService) Delete(id int) error {
	if err := service.EncounterRepo.Delete(id); err != nil {
		return fmt.Errorf("Can't find id to delete")
	}
	return nil
}

func (service *EncounterService) Update(encounter *model.Encounter) (*model.Encounter, error) {
	if err := service.EncounterRepo.Update(encounter); err == nil {
		return encounter, nil
	}

	return nil, fmt.Errorf("Couldn't approve.")
}
func (service *EncounterService) Approve(encounter *model.Encounter) (*model.Encounter, error) {

	encounter.ApprovalStatus = model.EncounterApprovalStatus("ADMIN_APPROVED")
	if err := service.EncounterRepo.Update(encounter); err == nil {
		return encounter, nil
	}

	return nil, fmt.Errorf("Couldn't approve.")

}

func (service *EncounterService) Decline(encounter *model.Encounter) (*model.Encounter, error) {

	encounter.ApprovalStatus = model.EncounterApprovalStatus("DECLINED")
	if err := service.EncounterRepo.Update(encounter); err == nil {
		return encounter, nil
	}

	return nil, fmt.Errorf("Couldn't approve.")
}

func findInVicinity(encountersPtr *[]*model.Encounter, lat, long float64) []*model.Encounter {
	encounters := *encountersPtr

	var results []*model.Encounter
	for _, e := range encounters {
		if distance := CalculateDistance(e.Latitude, e.Longitude, lat, long); distance <= e.Range {
			fmt.Println(distance, e.Range)
			results = append(results, e)
		}
	}
	return results
}

func getLatLong(userId int) (float64, float64, error) {
	url := "http://localhost:8000/positions/getByUser/" + strconv.Itoa(userId)

	response, err := http.Get(url)
	if err != nil {
		return 0, 0, err
	}
	defer response.Body.Close()

	body, err := io.ReadAll(response.Body)
	if err != nil {
		return 0, 0, err
	}

	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return 0, 0, err
	}

	long, ok1 := data["Longitude"].(float64)
	lat, ok2 := data["Latitude"].(float64)

	if !ok1 || !ok2 {
		return 0, 0, err
	}

	return lat, long, nil
}

File: BACK/src/Modules/Encounters/Explorer.Encounters.Core/UseCases/EncounterService.cs
using System.Text.Json;
using AutoMapper;
using Explorer.BuildingBlocks.Core.UseCases;
using Explorer.Encounters.API.Dtos;
using Explorer.Encounters.API.Public;
using Explorer.Encounters.Core.Domain;
using Explorer.Encounters.Core.Domain.Enums;
using Explorer.Encounters.Core.Domain.RepositoryInterfaces;
using Explorer.Stakeholders.API.Dtos;
using Explorer.Stakeholders.API.Internal;
using Explorer.Stakeholders.Core.Domain;
using Explorer.Tours.API.Internal;
using Explorer.Tours.API.Dtos;
using FluentResults;


namespace Explorer.Encounters.Core.UseCases
{
    public class EncounterService: CrudService<EncounterDto, Encounter>, IEncounterService
    {
        protected IEncounterRepository _encounterRepository;
        protected IInternalUserService _userService;
        protected IInternalProfileService _profileService;
        protected IInternalTouristPositionService _touristPositionService;

        public EncounterService(IEncounterRepository encounterRepository,
            IInternalUserService userService, IInternalProfileService profileService,
            IInternalTouristPositionService touristPositionService, IMapper mapper) : base(encounterRepository, mapper)
        {
            _encounterRepository = encounterRepository;
            _userService = userService;
            _profileService = profileService;
            _touristPositionService = touristPositionService;
        }

        public Result<PagedResult<EncounterDto>> GetApproved(int page, int pageSize)
        {
            var result = _encounterRepository.GetApproved(page, pageSize);
            return MapToDto(result);
        }

        public Result<PagedResult<EncounterDto>> GetApprovedByStatus(int page, int pageSize, string status)
        {
            if (Enum.TryParse<EncounterStatus>(status, out var encounterStatus))
            {
                var encounters = _encounterRepository.GetApprovedByStatus(page, pageSize, encounterStatus);
                return MapToDto(encounters);
            }
            else
            {
                return Result.Fail(FailureCode.InvalidArgument).WithError("Invalid status value");
            }
        }

        public Result<PagedResult<EncounterDto>> GetByUser(int page, int pageSize, long userId)
        {
            var result = _encounterRepository.GetByUser(page, pageSize, userId);
            return MapToDto(result);
        }

        public Result<PagedResult<EncounterDto>> GetTouristCreatedEncounters(int page, int pageSize)
        {
            var result = _encounterRepository.GetTouristCreatedEncounters(page, pageSize);
            return MapToDto(result);
        }

        public override Result<EncounterDto> Create(EncounterDto encounterDto)
        {
            try
            {
                UserDto user = _userService.Get(encounterDto.UserId).ValueOrDefault;
                if (user.Role == (int)UserRole.Tourist)
                {                 
                    if (_profileService.CanTouristCreateEncounters(encounterDto.UserId).Value) 
                    {
                        var encounter = MapToDomain(encounterDto);
                        encounter.UpdateApprovalStatus(EncounterApprovalStatus.PENDING);

                        var jsonEncounter = JsonSerializer.Serialize(encounterDto);

                        string url = "localhost:8083/encounters";
                        HttpClient client = new HttpClient();

                        client.DefaultRequestHeaders.Accept.Clear();
                        client.DefaultRequestHeaders.Accept.Add(
                            new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"));

                        HttpResponseMessage response = client.PostAsync("http://localhost:8083/encounters",
                            new StringContent(jsonEncounter, System.Text.Encoding.UTF8, "application/json")).Result;

                        if (response.IsSuccessStatusCode)
                        {
                            string json = response.Content.ReadAsStringAsync().Result;
                            EncounterDto result = JsonSerializer.Deserialize<EncounterDto>(json);

                            return Result.Ok(result);

                        }

                        return Result.Fail<EncounterDto>(response.ReasonPhrase); 

                    }
                    else
                    {
                        return Result.Fail("Encounter can not be created.");
                    }
                }
                else
                {
                    var encounter = MapToDomain(encounterDto);
                    encounter.UpdateApprovalStatus(EncounterApprovalStatus.PENDING);

                    var jsonEncounter = JsonSerializer.Serialize(encounterDto);

                    string url = "localhost:8083/encounters";
                    HttpClient client = new HttpClient();

                    client.DefaultRequestHeaders.Accept.Clear();
                    client.DefaultRequestHeaders.Accept.Add(
                        new System.Net.Http.Headers.MediaTypeWithQualityHeaderValue("application/json"));

                    HttpResponseMessage response = client.PostAsync("http://localhost:8083/encounters",
                        new StringContent(jsonEncounter, System.Text.Encoding.UTF8, "application/json")).Result;

                    if (response.IsSuccessStatusCode)
                    {
                        string json = response.Content.ReadAsStringAsync().Result;
                        EncounterDto result = JsonSerializer.Deserialize<EncounterDto>(json);

                        return Result.Ok(result);

                    }

                    return Result.Fail<EncounterDto>(response.ReasonPhrase); 
                }               
            }
            catch (Exception e)
            {
                return Result.Fail(FailureCode.InvalidArgument).WithError(e.Message);
            }
        }

        public Result<EncounterDto> Approve(EncounterDto encounterDto)
        {
            try
            {
                var encounter = MapToDomain(encounterDto);
                encounter.UpdateApprovalStatus(EncounterApprovalStatus.ADMIN_APPROVED);
                var result = _encounterRepository.Update(encounter);
                return MapToDto(result);
            }
            catch (Exception e)
            {
                return Result.Fail(FailureCode.InvalidArgument).WithError(e.Message);
            }
        }

        public Result<EncounterDto> Decline(EncounterDto encounterDto)
        {
            try
            {
                var encounter = MapToDomain(encounterDto);
                encounter.UpdateApprovalStatus(EncounterApprovalStatus.DECLINED);
                var result = _encounterRepository.Update(encounter);
                return MapToDto(result);
            }
            catch (Exception e)
            {
                return Result.Fail(FailureCode.InvalidArgument).WithError(e.Message);
            }
        }


        public Result<PagedResult<EncounterDto>> GetNearbyHidden(int page, int pageSize, int userId)
        {
            try {
                TouristPositionDto touristPosition = _touristPositionService.GetByUser(userId).Value;
                var encounters = _encounterRepository.GetNearbyByType(page, pageSize, touristPosition.Longitude, touristPosition.Latitude, EncounterType.LOCATION);
                return MapToDto(encounters);

            }catch(Exception e)
            {
                return Result.Fail(FailureCode.NotFound).WithError(e.Message);
            }
        }
        public Result<PagedResult<EncounterDto>> GetNearby(int page, int pageSize, int userId)
        {
            try {
                TouristPositionDto touristPosition = _touristPositionService.GetByUser(userId).Value;
                var encounters = _encounterRepository.GetNearby(page, pageSize, touristPosition.Longitude, touristPosition.Latitude);
                return MapToDto(encounters);

            }catch(Exception e)
            {
                return Result.Fail(FailureCode.NotFound).WithError(e.Message);
            }
        }
    }
}
